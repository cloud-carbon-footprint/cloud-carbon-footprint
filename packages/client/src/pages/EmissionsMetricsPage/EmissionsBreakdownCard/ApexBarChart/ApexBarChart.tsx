/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React, { FunctionComponent, useState } from 'react'
import Chart from 'react-apexcharts'
import { useTheme } from '@material-ui/core/styles'
import { PageEntry, Page, barChartCustomColors } from '../../../../Types'
import useStyles from './apexBarChartStyles'
import { createCustomBarColors, mapToRange } from './helpers'
import Pagination from '../Pagination'
import CarbonIntensityRange from '../CarbonIntensityRange/'
import { EmissionRatioResult } from '@cloud-carbon-footprint/common'

type ApexBarChartProps = {
  data: { string: [string, number] }
  dataType?: string
  emissionsData: EmissionRatioResult[]
}

const ApexBarChart: FunctionComponent<ApexBarChartProps> = ({
  data,
  dataType,
  emissionsData,
}) => {
  const [pageData, setPageData] = useState<Page<PageEntry>>({
    data: [],
    page: 0,
  })

  const theme = useTheme()
  const classes = useStyles()
  const mainTheme = theme.palette.primary.main
  const darkTheme = theme.palette.primary.dark

  let customBarColors = [mainTheme]
  if (dataType === 'region' && !!emissionsData.length) {
    customBarColors = createCustomBarColors(pageData, emissionsData, mainTheme)
  }

  const dataEntries: { x: string[]; y: number }[] = Object.entries(data)
    .filter((item: [string, [string, number]]) => item[1][1] > 0)
    .map((item) => {
      return {
        x: [item[0], `(${item[1][0]})`],
        y: item[1][1],
      }
    })
    .sort((higherC02, lowerCO2) => lowerCO2.y - higherC02.y)

  const smallestCO2E = dataEntries?.[dataEntries?.length - 1]?.y
  const largestCO2E = dataEntries?.[0]?.y
  const totalCO2EByDataType = dataEntries.reduce((acc, currentValue) => {
    return acc + currentValue.y
  }, 0)

  const pageSize = 10
  const minThreshold = 1
  const maxThreshold = 100
  const mappedDataEntries: PageEntry[] = dataEntries.map((entry) => {
    const yEntry = mapToRange(
      entry.y,
      smallestCO2E,
      largestCO2E,
      minThreshold,
      maxThreshold,
    )
    return { x: entry.x, y: isNaN(yEntry) ? maxThreshold : yEntry }
  })

  const toolbarOffset = {
    x: -120,
    y: dataType === 'region' ? -122 : -55,
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const options: any = {
    series: [
      {
        name: 'Total CO2e',
        data: pageData.data,
      },
    ],
    colors: customBarColors,
    chart: {
      type: 'bar',
      toolbar: {
        offsetX: toolbarOffset.x,
        offsetY: toolbarOffset.y,
        tools: {
          download: `
            <div class="apexcharts-menu-icon" title="Menu">
                <svg class="MuiSvgIcon-root" focusable="false" viewBox="0 0 24 24" aria-hidden="true">
                <path d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"></path>
                </svg>
            </div>   
           `,
        },
      },
    },
    grid: {
      show: false,
      yaxis: {
        lines: {
          show: false,
        },
      },
      xaxis: {
        lines: {
          show: false,
        },
      },
      padding: {
        left: dataType === 'region' ? -15 : -5,
      },
    },
    plotOptions: {
      bar: {
        horizontal: true,
        barHeight: `${7 * pageData.data.length}%`,
        distributed: true,
      },
    },
    legend: {
      show: false,
    },
    dataLabels: {
      enabled: true,
      textAnchor: 'start',
      formatter: function (_: number, opts: { dataPointIndex: number }) {
        const currentCO2E =
          dataEntries[pageData.page * pageSize + opts.dataPointIndex]?.y

        const formattedPercentage = (currentCO2E / totalCO2EByDataType) * 100
        return formattedPercentage < 0.01
          ? '< 0.01 %'
          : `${formattedPercentage.toFixed(2)} %`
      },
      offsetX: 16,
      background: {
        enabled: true,
        foreColor: mainTheme,
        borderColor: darkTheme,
        padding: 6,
        borderRadius: 1,
        borderWidth: 1,
        opacity: 0.9,
      },
    },
    xaxis: {
      type: 'category',
      labels: {
        style: {
          fontSize: 0,
        },
      },
      axisBorder: {
        show: false,
      },
      max: maxThreshold,
    },
    yaxis: {
      labels: {
        minWidth: 150,
        maxWidth: 150,
        offsetY: 5,
        style: {
          fontSize: '12px',
        },
      },
    },
    tooltip: {
      fillSeriesColor: false,
      y: {
        formatter: function (value: number, opts: { dataPointIndex: number }) {
          return `${dataEntries[
            pageData.page * pageSize + opts.dataPointIndex
          ].y.toFixed(3)} metric tons`
        },
      },
    },
    height: '500px',
  }

  const handlePage = (page: Page<PageEntry>) => {
    setPageData(page)
  }

  return (
    <div className={classes.barChartContainer}>
      {dataType === 'region' && (
        <CarbonIntensityRange
          startLabel="Low carbon intensity"
          endLabel="High carbon intensity"
          colorRange={barChartCustomColors}
        />
      )}
      <Chart
        options={options}
        series={options.series}
        type="bar"
        height={options.height}
      />
      <Pagination
        data={mappedDataEntries}
        pageSize={pageSize}
        handlePage={handlePage}
      />
    </div>
  )
}

export default ApexBarChart
