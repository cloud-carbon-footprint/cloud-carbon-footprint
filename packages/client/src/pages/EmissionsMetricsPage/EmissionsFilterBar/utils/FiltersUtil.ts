/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { DropdownFilterOptions, DropdownOption, FilterOptions } from 'Types'
import createOptionChooser from './options'

export type DropdownSelections = {
  [key in DropdownFilterOptions]: DropdownOption[]
}

export function handleDropdownSelections(
  filterType: DropdownFilterOptions,
  selections: DropdownOption[],
  oldSelections: DropdownSelections,
  filterOptions: FilterOptions,
): DropdownSelections {
  return createOptionChooser(
    filterType,
    selections,
    oldSelections,
    filterOptions,
  ).choose()
}

export function numSelectedLabel(
  length: number,
  totalLength: number,
  type = 'Services',
): string {
  const lengthWithoutAllOption = totalLength - 1
  if (length === totalLength) {
    return `${type}: ${lengthWithoutAllOption} of ${lengthWithoutAllOption}`
  } else {
    return `${type}: ${length} of ${lengthWithoutAllOption}`
  }
}
