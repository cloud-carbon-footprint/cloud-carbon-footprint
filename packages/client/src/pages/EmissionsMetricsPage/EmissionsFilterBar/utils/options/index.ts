/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { DropdownOption, FilterOptions, DropdownFilterOptions } from 'Types'
import { OptionChooser } from './OptionChooser'
import { DropdownSelections } from '../FiltersUtil'
import { CloudProviderChooser } from './CloudProviderChooser'
import { AccountChooser } from './AccountChooser'
import { ServiceChooser } from './ServiceChooser'

export default function createOptionChooser(
  filterType: DropdownFilterOptions,
  selections: DropdownOption[],
  oldSelections: DropdownSelections,
  filterOptions: FilterOptions,
): OptionChooser {
  switch (filterType) {
    case DropdownFilterOptions.CLOUD_PROVIDERS:
      return new CloudProviderChooser(selections, oldSelections, filterOptions)
    case DropdownFilterOptions.ACCOUNTS:
      return new AccountChooser(selections, oldSelections, filterOptions)
    case DropdownFilterOptions.SERVICES:
      return new ServiceChooser(selections, oldSelections, filterOptions)
  }
}
