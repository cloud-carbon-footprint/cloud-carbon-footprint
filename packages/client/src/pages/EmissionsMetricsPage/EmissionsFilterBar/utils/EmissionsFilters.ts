/*
 * Â© 2021 Thoughtworks, Inc.
 */

import moment from 'moment'
import { Filters } from '../../../../common/FilterBar/utils/Filters'
import {
  DropdownFilterOptions,
  DropdownOption,
  FilterOptions,
  FilterResultResponse,
  FiltersConfig,
  unknownOptionTypes,
} from '../../../../Types'
import {
  ALL_ACCOUNTS_DROPDOWN_OPTION,
  ALL_DROPDOWN_FILTER_OPTIONS,
  ALL_SERVICES_DROPDOWN_OPTION,
  CLOUD_PROVIDER_OPTIONS,
} from '../../../../common/FilterBar/utils/DropdownConstants'
import { EstimationResult, ServiceData } from '@cloud-carbon-footprint/common'
import { AccountChooser } from './options/AccountChooser'
import { DropdownSelections } from '../../../../common/FilterBar/utils/FiltersUtil'
import { OptionChooser } from '../../../../common/FilterBar/utils/OptionChooser'
import { CloudProviderChooser } from './options/CloudProviderChooser'
import { ServiceChooser } from './options/ServiceChooser'

const defaultConfig: FiltersConfig = {
  timeframe: 36,
  dateRange: null,
  options: {
    [DropdownFilterOptions.SERVICES]: [ALL_SERVICES_DROPDOWN_OPTION],
    [DropdownFilterOptions.CLOUD_PROVIDERS]: CLOUD_PROVIDER_OPTIONS,
    [DropdownFilterOptions.ACCOUNTS]: [ALL_ACCOUNTS_DROPDOWN_OPTION],
  },
}

export class EmissionsFilters extends Filters {
  constructor(config = defaultConfig) {
    super(config)
  }

  static generateConfig(newOptions: FilterResultResponse): FiltersConfig {
    return super.generateConfig(newOptions, defaultConfig)
  }

  create(config: FiltersConfig): Filters {
    return new EmissionsFilters(config)
  }

  createOptionChooser(
    filterType: DropdownFilterOptions,
    selections: DropdownOption[],
    oldSelections: DropdownSelections,
    filterOptions: FilterOptions,
  ): OptionChooser {
    switch (filterType) {
      case DropdownFilterOptions.CLOUD_PROVIDERS:
        return new CloudProviderChooser(
          selections,
          oldSelections,
          filterOptions,
        )
      case DropdownFilterOptions.ACCOUNTS:
        return new AccountChooser(selections, oldSelections, filterOptions)
      case DropdownFilterOptions.SERVICES:
        return new ServiceChooser(selections, oldSelections, filterOptions)
    }
  }

  filter(rawResults: EstimationResult[]): EstimationResult[] {
    const resultsFilteredByTime = this.getResultsFilteredByTime(rawResults)
    const resultsFilteredByService = this.getResultsFilteredBy(
      DropdownFilterOptions.SERVICES,
      resultsFilteredByTime,
    )
    return this.getResultsFilteredBy(
      DropdownFilterOptions.ACCOUNTS,
      resultsFilteredByService,
    ).filter((estimationResult) => !!estimationResult?.serviceEstimates?.length)
  }

  private getResultsFilteredByTime(
    previousResults: EstimationResult[],
  ): EstimationResult[] {
    const today = moment.utc()
    let start: moment.Moment
    let end: moment.Moment
    if (this.timeframe < 0 && this.dateRange?.startDate) {
      start = this.dateRange.startDate
      end = this.dateRange.endDate || today
    } else {
      end = today
      start = end.clone().subtract(this.timeframe, 'M')
    }

    return previousResults.filter((estimationResult: EstimationResult) =>
      moment.utc(estimationResult.timestamp).isBetween(start, end, 'day', '[]'),
    )
  }

  private getResultsFilteredBy(
    type: string,
    previousResults: EstimationResult[],
  ): EstimationResult[] {
    const hasAllOptionsSelected = this.options[type].includes(
      ALL_DROPDOWN_FILTER_OPTIONS[type],
    )
    return previousResults.map((estimationResult) => {
      const filteredServiceEstimates = estimationResult.serviceEstimates.filter(
        (serviceEstimate) =>
          this.options[type].some((dropdownOption) =>
            this.isUnknownOrSameName(dropdownOption, type, serviceEstimate),
          ) || hasAllOptionsSelected,
      )
      return {
        timestamp: estimationResult.timestamp,
        serviceEstimates: filteredServiceEstimates,
        periodStartDate: estimationResult.periodStartDate,
        periodEndDate: estimationResult.periodEndDate,
        groupBy: estimationResult.groupBy,
      }
    })
  }

  private isUnknownOrSameName(
    dropdownOption: DropdownOption,
    type: string,
    serviceEstimate: ServiceData,
  ) {
    const name =
      type === DropdownFilterOptions.ACCOUNTS ? 'accountName' : 'serviceName'
    return (
      (dropdownOption.name.includes(unknownOptionTypes[type]) &&
        serviceEstimate[name] === null) ||
      dropdownOption.name === serviceEstimate[name]
    )
  }
}
