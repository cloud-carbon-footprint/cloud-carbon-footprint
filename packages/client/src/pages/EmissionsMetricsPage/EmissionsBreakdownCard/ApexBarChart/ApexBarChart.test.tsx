/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React from 'react'
import { act, create, ReactTestRenderer } from 'react-test-renderer'
import Chart from 'react-apexcharts'
import moment from 'moment'
import {
  EmissionRatioResult,
  EstimationResult,
  GroupBy,
} from '@cloud-carbon-footprint/common'
import { fakeEmissionFactors } from '../../../../utils/data'
import { sumCO2ByServiceOrRegion } from '../../../../utils/helpers'
import { Page, PageEntry } from '../../../../Types'
import ApexBarChart from './ApexBarChart'
import Pagination from '../Pagination'
import { createCustomBarColors } from './helpers'

jest.mock('apexcharts')

describe('ApexBarChart', () => {
  let fixture: ReactTestRenderer
  let barChartData: { string: [string, number] }
  const timestamp: Date = moment('2019-08-10T00:00:00.000Z').toDate()
  const data: EstimationResult[] = [
    {
      timestamp: timestamp,
      periodStartDate: timestamp,
      periodEndDate: timestamp,
      groupBy: GroupBy.day,
      serviceEstimates: [
        {
          cloudProvider: 'AWS',
          accountId: 'some account id',
          accountName: 'some account',
          serviceName: 'ebs',
          kilowattHours: 0,
          co2e: 3000.014,
          cost: 0,
          region: 'us-west-1',
        },
        {
          cloudProvider: 'AWS',
          accountId: 'some account id',
          accountName: 'some account',
          serviceName: 's3',
          kilowattHours: 0,
          co2e: 1000.014,
          cost: 0,
          region: 'us-west-2',
        },
        {
          cloudProvider: 'AWS',
          accountId: 'some account id',
          accountName: 'some account',
          serviceName: 'ec2',
          kilowattHours: 0,
          co2e: 2000.014,
          cost: 0,
          region: 'us-west-3',
        },
        {
          cloudProvider: 'AWS',
          accountId: 'some account id',
          accountName: 'some account',
          serviceName: 'eks',
          kilowattHours: 0,
          co2e: 0.000014,
          cost: 0,
          region: 'us-west-4',
        },
      ],
    },
  ]
  beforeEach(() => {
    barChartData = sumCO2ByServiceOrRegion(
      data as EstimationResult[],
      'service',
    )
    fixture = create(
      <ApexBarChart
        data={barChartData}
        dataType="service"
        emissionsData={fakeEmissionFactors}
      />,
    )
  })

  afterEach(() => {
    fixture.unmount()
  })

  it('renders with correct configuration', () => {
    expect(fixture.toJSON()).toMatchSnapshot()
  })

  it('should format tool tip values with proper data instead of scaled down data', () => {
    const handlePage: (page: Page<PageEntry>) => void =
      fixture.root.findByType(Pagination).props?.handlePage
    // make pagination send first page
    act(() => {
      handlePage({
        data: [
          { x: ['ebs', '(AWS)'], y: 100 },
          { x: ['ec2', '(AWS)'], y: 67.00015384528277 },
          { x: ['s3', '(AWS)'], y: 34.00030769056555 },
          { x: ['eks', '(AWS)'], y: 1 },
        ],
        page: 0,
      })
    })

    const yFormatter =
      fixture.root.findByType(Chart).props?.options?.tooltip?.y?.formatter
    expect(yFormatter).toBeDefined()
    expect(yFormatter(null, { dataPointIndex: 1 })).toEqual(
      '2000.014 metric tons',
    )
  })

  it('should format data label values with proper data instead of scaled down data', () => {
    const handlePage: (page: Page<PageEntry>) => void =
      fixture.root.findByType(Pagination).props?.handlePage
    // make pagination send first page
    act(() => {
      handlePage({
        data: [
          { x: ['ebs', '(AWS)'], y: 100 },
          { x: ['ec2', '(AWS)'], y: 67.00015384528277 },
          { x: ['s3', '(AWS)'], y: 34.00030769056555 },
          { x: ['eks', '(AWS)'], y: 1 },
        ],
        page: 0,
      })
    })

    const dataLabelFormatter =
      fixture.root.findByType(Chart).props?.options?.dataLabels?.formatter
    expect(dataLabelFormatter).toBeDefined()
    expect(dataLabelFormatter(null, { dataPointIndex: 1 })).toEqual('33.33 %')
  })

  it('should format data label values that are less than 0.01', () => {
    const handlePage: (page: Page<PageEntry>) => void =
      fixture.root.findByType(Pagination).props?.handlePage
    // make pagination send first page
    act(() => {
      handlePage({
        data: [
          { x: ['ebs', '(AWS)'], y: 100 },
          { x: ['ec2', '(AWS)'], y: 67.00015384528277 },
          { x: ['s3', '(AWS)'], y: 34.00030769056555 },
          { x: ['eks', '(AWS)'], y: 1 },
        ],
        page: 0,
      })
    })

    const dataLabelFormatter =
      fixture.root.findByType(Chart).props?.options?.dataLabels?.formatter
    expect(dataLabelFormatter).toBeDefined()
    expect(dataLabelFormatter(null, { dataPointIndex: 3 })).toEqual('< 0.01 %')
  })

  it('should filter, sort, order, and scale down data before passing it to Pagination component', function () {
    const paginationComponent = fixture.root.findByType(Pagination)
    const sortedData = [
      { x: ['ebs', '(AWS)'], y: 100 },
      { x: ['ec2', '(AWS)'], y: 67.00015384528277 },
      { x: ['s3', '(AWS)'], y: 34.00030769056555 },
      { x: ['eks', '(AWS)'], y: 1 },
    ]

    expect(paginationComponent.props.data).toEqual(sortedData)
  })

  it('should set colors to each bar based on region emissions', () => {
    barChartData = sumCO2ByServiceOrRegion(data as EstimationResult[], 'region')
    fixture = create(
      <ApexBarChart
        data={barChartData}
        dataType="region"
        emissionsData={fakeEmissionFactors}
      />,
    )
    expect(fixture.toJSON()).toMatchSnapshot()
    const paginationComponent = fixture.root.findByType(Pagination)
    const sortedData = [
      { x: ['us-west-1', '(AWS)'], y: 100 },
      { x: ['us-west-3', '(AWS)'], y: 67.00015384528277 },
      { x: ['us-west-2', '(AWS)'], y: 34.00030769056555 },
      { x: ['us-west-4', '(AWS)'], y: 1 },
    ]

    const customColors = ['#790000', '#D99200', '#DF5200', '#00791E']

    expect(paginationComponent.props.data).toEqual(sortedData)

    const optionColors = fixture.root.findByType(Chart).props?.options?.colors
    expect(optionColors).toBeDefined()
    expect(optionColors).toEqual(customColors)
  })

  it('should create custom colors array', () => {
    const firstPagedata = [
      { x: ['us-west-1', '(AWS)'], y: 100 },
      { x: ['us-west-3', '(AWS)'], y: 67.00015384528277 },
      { x: ['us-west-2', '(AWS)'], y: 34.00030769056555 },
      { x: ['us-west-4', '(AWS)'], y: 1 },
      { x: ['us-east-1', '(AWS)'], y: 2 },
    ]
    const pageData: Page<PageEntry> = { data: firstPagedata, page: 0 }
    const mainTheme = '#2C82BE'

    const emissionsData: EmissionRatioResult[] = fakeEmissionFactors
    const colors = ['#790000', '#D99200', '#DF5200', '#00791E', '#73B500']
    const defaultColors = [
      '#2C82BE',
      '#2C82BE',
      '#2C82BE',
      '#2C82BE',
      '#2C82BE',
    ]

    const customColors = createCustomBarColors(
      pageData,
      emissionsData,
      mainTheme,
    )

    expect(customColors).toEqual(colors)

    const mainColors = createCustomBarColors(pageData, [], mainTheme)

    expect(mainColors).toEqual(defaultColors)
  })
})
