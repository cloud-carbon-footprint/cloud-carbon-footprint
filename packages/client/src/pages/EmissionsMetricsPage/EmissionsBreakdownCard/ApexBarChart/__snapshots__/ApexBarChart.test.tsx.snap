// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApexBarChart renders with correct configuration 1`] = `
<div
  className="makeStyles-barChartContainer-1"
>
  <div />
  <div
    className="makeStyles-paginationContainer-2"
  >
    <span
      className="makeStyles-paginationLabel-3"
    >
      1 - 4
       of 
      4
    </span>
    <button
      aria-label="first page"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="chevron-left"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="chevron-right"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="last page"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`ApexBarChart should set colors to each bar based on region emissions 1`] = `
<div
  className="makeStyles-barChartContainer-16"
>
  <div
    style={
      Object {
        "alignSelf": "center",
        "display": "flex",
        "flexDirection": "column",
        "padding": "10px 45px",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "fontWeight": 500,
          "justifyContent": "space-between",
          "paddingBottom": "12px",
        }
      }
    >
      <div>
        Low carbon intensity
      </div>
      <div>
        High carbon intensity
      </div>
    </div>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <div
        className="t-dot"
        data-testid="timeline-dot"
        style={
          Object {
            "backgroundColor": "#73B500",
            "borderRadius": "8px",
            "height": "15px",
            "width": "15px",
          }
        }
      />
      <div
        className="t-bar"
        style={
          Object {
            "backgroundColor": "#aaaaaa",
            "flexGrow": 2,
            "height": "1px",
          }
        }
      />
      <div
        className="t-dot"
        data-testid="timeline-dot"
        style={
          Object {
            "backgroundColor": "#00791E",
            "borderRadius": "8px",
            "height": "15px",
            "width": "15px",
          }
        }
      />
      <div
        className="t-bar"
        style={
          Object {
            "backgroundColor": "#aaaaaa",
            "flexGrow": 2,
            "height": "1px",
          }
        }
      />
      <div
        className="t-dot"
        data-testid="timeline-dot"
        style={
          Object {
            "backgroundColor": "#D99200",
            "borderRadius": "8px",
            "height": "15px",
            "width": "15px",
          }
        }
      />
      <div
        className="t-bar"
        style={
          Object {
            "backgroundColor": "#aaaaaa",
            "flexGrow": 2,
            "height": "1px",
          }
        }
      />
      <div
        className="t-dot"
        data-testid="timeline-dot"
        style={
          Object {
            "backgroundColor": "#DF5200",
            "borderRadius": "8px",
            "height": "15px",
            "width": "15px",
          }
        }
      />
      <div
        className="t-bar"
        style={
          Object {
            "backgroundColor": "#aaaaaa",
            "flexGrow": 2,
            "height": "1px",
          }
        }
      />
      <div
        className="t-dot"
        data-testid="timeline-dot"
        style={
          Object {
            "backgroundColor": "#790000",
            "borderRadius": "8px",
            "height": "15px",
            "width": "15px",
          }
        }
      />
    </div>
  </div>
  <div />
  <div
    className="makeStyles-paginationContainer-17"
  >
    <span
      className="makeStyles-paginationLabel-18"
    >
      1 - 4
       of 
      4
    </span>
    <button
      aria-label="first page"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="chevron-left"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="chevron-right"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="last page"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
